dali/pipeline/operators/python_function/python_function.cc

```diff
@@ -0,0 +1,120 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/python_function/python_function.h"
+#include <vector>
+
+namespace dali {
+
+DALI_SCHEMA(PythonFunction)
+        .DocStr(R"code(Executes python function that consumes and produces single numpy array.)code")
+        .NumInput(1)
+        .NumOutput(1)
+        .AllowMultipleInputSets()
+        .AddArg("function_id",
+                R"code(Id of the python function.)code",
+                DALI_INT64);
+
+struct PyBindInitializer {
+  PyBindInitializer() {
+    auto thread_state = PyGILState_Ensure();
+    pybind11::get_shared_data("");  // setup the pybind's internals pointer
+    PyGILState_Release(thread_state);
+  }
+};
+
+// Some pybind's internals are not properly initialized when used in dynamically linked library,
+// so this workaround initializes them manually
+static PyBindInitializer pybind_initializer{}; // NOLINT
+
+py::array TensorToNumpyArray(const Tensor<CPUBackend> &tensor) {
+  DALI_TYPE_SWITCH(tensor.type().id(), DType,
+    return py::array_t<DType, py::array::c_style>(tensor.shape(), tensor.template data<DType>());
+  )
+}
+
+TypeInfo GetDaliType(const py::array &array) {
+  switch (array.dtype().kind()) {
+    case 'i': {
+      switch (array.itemsize()) {
+        case sizeof(int16):
+          return TypeInfo::Create<int16>();
+        case sizeof(int32):
+          return TypeInfo::Create<int32>();
+        case sizeof(int64):
+          return TypeInfo::Create<int64>();
+        default: {}
+      }
+      break;
+    }
+    case 'u': {
+      switch (array.itemsize()) {
+        case sizeof(uint8):
+          return TypeInfo::Create<uint8>();
+        default: {}
+      }
+      break;
+    }
+    case 'f': {
+      switch (array.itemsize()) {
+        case sizeof(float16):
+          return TypeInfo::Create<float16>();
+        case sizeof(float):
+          return TypeInfo::Create<float>();
+        case sizeof(double):
+          return TypeInfo::Create<double>();
+        default: {}
+      }
+      break;
+    }
+    case 'b': {
+      return TypeInfo::Create<bool>();
+    }
+    default: {}
+  }
+  DALI_FAIL(
+      "Unexpected numpy array type: " + array.dtype().kind() + std::to_string(array.itemsize()));
+}
+
+py::array NumpyArrayAsContiguous(const TypeInfo &type, const py::array &array) {
+  DALI_TYPE_SWITCH(type.id(), DType,
+    return py::array_t<DType, py::array::c_style>::ensure(array);
+  )
+}
+
+void CopyNumpyArrayToTensor(Tensor<CPUBackend> &tensor, const py::array &array) {
+  std::vector<Index> shape(static_cast<size_t>(array.ndim()));
+  std::copy(array.shape(), array.shape() + array.ndim(), shape.begin());
```

Comment:
This function requires unit-testing to verify it actually works

Reply:
Writing unit tests for this single function would require linking the test binary with python libs and to do that i would need to add them to dependencies and it's not trivial to do that for all supported versions of python. I can do that but I'm not sure if the profit/effort ratio is big enough.

Reply:
If this function is covered by python test, then it should be enough as long as CI runs it in both 2.7 and 3.x.

Reply:
Having given it a thought - it may be beneficial to add a test in Python aimed at this particular function (name of the test should imply that). The test would check whether contiguous/non-contiguous arrays are processed correctly in a pass-through pipeline.